pkg/authorization/api/helpers.go

```diff
@@ -244,6 +244,59 @@ func SubjectsStrings(currentNamespace string, subjects []kapi.ObjectReference) (
 	return users, groups, sas, others
 }
 
+// SubjectsContainUser returns true if the provided subjects contain the named user. currentNamespace
+// is used to identify service accounts that are defined in a relative fashion.
+func SubjectsContainUser(subjects []kapi.ObjectReference, currentNamespace string, user string) bool {
+	if !strings.HasPrefix(user, serviceaccount.ServiceAccountUsernamePrefix) {
+		for _, subject := range subjects {
+			switch subject.Kind {
+			case UserKind, SystemUserKind:
+				if user == subject.Name {
+					return true
+				}
+			}
+		}
+		return false
+	}
+
+	for _, subject := range subjects {
+		switch subject.Kind {
+		case ServiceAccountKind:
+			namespace := currentNamespace
+			if len(subject.Namespace) > 0 {
+				namespace = subject.Namespace
+			}
+			if len(namespace) == 0 {
+				continue
+			}
+			if user == serviceaccount.MakeUsername(namespace, subject.Name) {
+				return true
+			}
+
+		case UserKind, SystemUserKind:
+			if user == subject.Name {
```

Comment:
Since a service account *is* a user, it seems more appropriate here to
explicitly check for User.  What's the alternative path?

On Tue, Jul 19, 2016 at 8:08 AM, David Eads <notifications@github.com>
wrote:

> In pkg/authorization/api/helpers.go
> <https://github.com/openshift/origin/pull/9836#discussion_r71325075>:
>
> > +	for _, subject := range subjects {
> > +		switch subject.Kind {
> > +		case ServiceAccountKind:
> > +			namespace := currentNamespace
> > +			if len(subject.Namespace) > 0 {
> > +				namespace = subject.Namespace
> > +			}
> > +			if len(namespace) == 0 {
> > +				continue
> > +			}
> > +			if user == serviceaccount.MakeUsername(namespace, subject.Name) {
> > +				return true
> > +			}
> > +
> > +		case UserKind, SystemUserKind:
> > +			if user == subject.Name {
>
> Conversion should have coerced everything shaped that way into a service
> account subject. Is that not happening?
>
> Ah, might allow them as a system user, but never a regular user.
> https://github.com/openshift/origin/blob/master/pkg/authorization/api/validation/validation.go#L304
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/openshift/origin/pull/9836/files/360a36e096aba8109ab1bdf19273e5b14805ade0#r71325075>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABG_pzbwfnX1RsgIrAuTBoBt_kHFeRdHks5qXL5UgaJpZM4JMAGj>
> .
>